resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocatorBlacklistTracker.scala

```diff
@@ -0,0 +1,155 @@
+
+  private val IS_YARN_ALLOCATION_BLACKLIST_ENABLED =
+    sparkConf.get(YARN_ALLOCATION_BLACKLIST_ENABLED).getOrElse(false)
+
+  private val BLACKLIST_MAX_FAILED_EXEC_PER_NODE = sparkConf.get(MAX_FAILED_EXEC_PER_NODE)
+
+  private val BLACKLIST_SIZE_LIMIT = sparkConf.get(YARN_BLACKLIST_SIZE_LIMIT)
+
+  private val BLACKLIST_SIZE_DEFAULT_WEIGHT = sparkConf.get(YARN_BLACKLIST_SIZE_DEFAULT_WEIGHT)
+
+  private var clock: Clock = new SystemClock
+
+  private val allocationBlacklistedNodesWithExpiry = new HashMap[String, Long]()
+
+  private var currentBlacklistedYarnNodes = Set.empty[String]
+
+  private var schedulerBlacklistedNodesWithExpiry = Map.empty[String, Long]
+
+  private var numClusterNodes = (Int.MaxValue / BLACKLIST_SIZE_DEFAULT_WEIGHT).toInt
+
+  def setNumClusterNodes(numClusterNodes: Int): Unit = {
+    this.numClusterNodes = numClusterNodes
+  }
+
+  /**
+   * Use a different clock. This is mainly used for testing.
+   */
+  def setClock(newClock: Clock): Unit = {
+    clock = newClock
+  }
+
+  def handleResourceAllocationFailure(hostOpt: Option[String]): Unit = {
+    hostOpt match {
+      case Some(hostname) =>
+        // failures on a already blacklisted nodes are not even tracked
+        // otherwise such failures could shutdown the application
+        // as resource requests are asynchronous
+        // and a late failure response could exceed MAX_EXECUTOR_FAILURES
+        if (!schedulerBlacklistedNodesWithExpiry.contains(hostname) &&
+          !allocationBlacklistedNodesWithExpiry.contains(hostname)) {
+          failureWithinTimeIntervalTracker.registerFailureOnHost(hostname)
+          updateAllocationBlacklistedNodes(hostname)
+        }
+      case None =>
+        failureWithinTimeIntervalTracker.registerSumExecutorFailure()
+    }
+  }
+
+  private def updateAllocationBlacklistedNodes(hostname: String): Unit = {
+    if (IS_YARN_ALLOCATION_BLACKLIST_ENABLED) {
+      val failuresOnHost = failureWithinTimeIntervalTracker.getNumExecutorFailuresOnHost(hostname)
+      if (failuresOnHost > BLACKLIST_MAX_FAILED_EXEC_PER_NODE) {
+        logInfo("blacklisting host as YARN allocation failed: %s".format(hostname))
+        allocationBlacklistedNodesWithExpiry.put(
+          hostname,
+          clock.getTimeMillis() + BLACKLIST_TIMEOUT_MILLIS)
+        refreshBlacklistedNodes()
+      }
+    }
+  }
+
+  def setSchedulerBlacklistedNodes(schedulerBlacklistedNodesWithExpiry: Map[String, Long]): Unit = {
+    this.schedulerBlacklistedNodesWithExpiry = schedulerBlacklistedNodesWithExpiry
+    refreshBlacklistedNodes()
+  }
+
+  private def refreshBlacklistedNodes(): Unit = {
+    removeExpiredYarnBlacklistedNodes()
+    val limit =
+      BLACKLIST_SIZE_LIMIT.getOrElse((numClusterNodes * BLACKLIST_SIZE_DEFAULT_WEIGHT).toInt)
+    val nodesToBlacklist =
+      if (schedulerBlacklistedNodesWithExpiry.size +
+        allocationBlacklistedNodesWithExpiry.size > limit) {
```

Comment:
nit: double-indent the continued continuation of the `if` condition.  (we dont' do this everywhere but we should, I find it helps)